### ISSUE TITLE
Investigate HEST_OS_study errors

### ISSUE THREAD

**Issue by rebkwok** [opensafely-core/cohort-extractor#716]
The `covid_admission_primary_diagnosis` variable in this (study definition)[https://github.com/opensafely/HEST_OS_study/blob/main/analysis/study_definition.py] is sometimes returning no data which causes (this)[https://github.com/opensafely/HEST_OS_study/blob/main/analysis/full_table_script.R#L91] to fail
```
covid_admission_primary_diagnosis=patients.admitted_to_hospital(
        returning="primary_diagnosis",
        with_these_diagnoses=covid_codelist,
        on_or_after= index_date,
        find_first_match_in_period=True,  
        date_format="YYYY-MM-DD", 
        return_expectations={"date": {"earliest": index_date},"incidence" : 0.25,
            "category": {"ratios": {"U071":0.5, "U072":0.5}},
        },
    ),```

The last run on TPP failed https://github.com/opensafely/HEST_OS_study/blob/main/analysis/full_table_script.R#L91

(Note the use of the `date_format` kwarg when the returning value is a string, not a date - this appears to be a red herring;  the date_format kwarg will be ignored)

**Comment by rebkwok** [comment]
I've tried running the SQL that is generated from the study definition for the specified population and covid_admission_primary_diagnosis only.  It produces results, but all the covid_admission_primary_diagnosis column is empty, and there is an error message
The invalid escape character "\\" was specified in a LIKE predicate
This is coming from the covid_admission_primary_diagnosis variable, which generates:
-- Query for covid_admission_primary_diagnosis
SELECT * INTO #covid_admission_primary_diagnosis FROM (
    SELECT
    t.Patient_ID AS patient_id,
    t.primary_diagnosis AS primary_diagnosis
    FROM (
        SELECT APCS.Patient_ID, Spell_Primary_Diagnosis AS primary_diagnosis,
        ROW_NUMBER() OVER (
            PARTITION BY APCS.Patient_ID
            ORDER BY Admission_Date ASC, APCS.APCS_Ident
            ) AS rownum
            FROM APCS
            INNER JOIN APCS_Der
            ON APCS.APCS_Ident = APCS_Der.APCS_Ident
    WHERE Admission_Date >= '20210514' AND (Der_Diagnosis_All LIKE '%[^A-Za-z0-9]U071%' ESCAPE '\\' OR Der_Diagnosis_All LIKE '%[^A-Za-z0-9]U072%' ESCAPE '\\')
    ) t
    WHERE rownum = 1
) t
If I remove the ESCAPE '\\' from the SQL, it runs successfully and does retrieve values for covid_admission_primary_diagnosis

**Comment by rebkwok** [comment]
From @evansd:
> our escaping for LIKE queries is broken, but we've never noticed because it isn't actually tested and we don't (yet) have any problematic characters in our codelists. Backslash is a unfortunate character to use here because there are multiple layers of string escaping going on (python strings->sql strings->LIKE expressions) and it all gets very confusing. I think we could swap this out for ! and life would be much easier.
That would involve changing this line from f"\\{char}" to f"!{char}":
https://github.com/opensafely-core/cohort-extractor/blob/a8b32fbdfe4754101d5624cad9f6d81deb209552/cohortextractor/tpp_backend.py#L2854
And then replacing all instances of ESCAPE '\\' with ESCAPE '!' e.g:
https://github.com/opensafely-core/cohort-extractor/blob/a8b32fbdfe4754101d5624cad9f6d81deb209552/cohortextractor/tpp_backend.py#L2347
The relevant MSSQL docs are here if that's helpful
https://docs.microsoft.com/en-us/sql/t-sql/language-elements/like-transact-sql?view=sql-server-ver15#pattern-matching-with-the-escape-clause
It would also be good to actually test that this escaping works somehow.

**Comment by rebkwok** [comment]
Also from @evansd 
> As soon as I hear "indeterminacy" my immediate thought is that the ORDER BY clause used in the ROW_NUMBER() OVER (PARTITION BY ... subquery doesn't have enough columns to give a fully determined ordering i.e. there are multiple rows with the same Admission_Date and APCS.APCS_Ident and therefore there isn't a uniquely determined first row. MSSQL's behaviour under these circumstances is to return a row (seemingly) at random so results will vary between the same query run against the same data.

**Comment by rebkwok** [comment]
This is still baffling me.  Jennifer says the intermittent failures (where the R script encounters columns with no data) are happening locally, which means it's to do with the generated dummy data, not (or at least not only) with the actual SQL run on the DB.  However, the expectations for the  `covid_admission_primary_diagnosis` variable have an incidence of 0.25, so with generated dummy data, there should always be values in the column.  I can't get this to fail locally.

In the actual outputs on the server from the last (failed) job, the logs do indeed say that the `decriptive_tables` job failed because there were no values for `covid_admission_primary_diagnosis`.  However, the generated input.csv file does have values.

**Comment by evansd** [comment]
This is really weird. I wonder if the problem is actually in the R script. The ideal thing would be if, the next time it fails locally, she could make a copy of the entire project directory, zip it up and send it to us.

> However, the generated input.csv file does have values.

Is it possible that it's been regenerated since the failing job was run? It should be possible to determine that by checking the job history. If not, then that definitely points to an issue with the R script (or possibly something more fundamental with job-runner/docker).

**Comment by rebkwok** [comment]
>  I wonder if the problem is actually in the R script.
I wondered that too, but I can't get that to fail either (except by actually removing all the values from the input.csv column)

> Is it possible that it's been regenerated since the failing job was run?
Maybe.  I looked at the file dates, but not the job history specifically.

**Comment by evansd** [comment]
Oh, file dates should be at least roughly correct so sounds like you've covered that already.

**Comment by rebkwok** [comment]
> As soon as I hear "indeterminacy" my immediate thought is that the ORDER BY clause used in the ROW_NUMBER() OVER (PARTITION BY ... subquery doesn't have enough columns to give a fully determined ordering i.e. there are multiple rows with the same Admission_Date and APCS.APCS_Ident and therefore there isn't a uniquely determined first row. MSSQL's behaviour under these circumstances is to return a row (seemingly) at random so results will vary between the same query run against the same data.

There's a comment in the code about the APCS_Ident probably being unique enough for the sorting.  But I've also checked the database, and for the dates this study is looking at (admission >= 2021-05-14), there are no duplicates of Patient_ID, Admission_Date and APCS.APCS_Ident.  But then, I'm not sure the intermittent-ness is anything to do with this anyway, if it's actually a local issue.

**Comment by evansd** [comment]
Right, sounds like I was on the wrong track completely there then. If it happens locally then, as you say, it can't be anything to do with the db.

**Comment by rebkwok** [comment]
It fails (or failed on 24th Nov at least) on TPP, but I don't know that it fails intermittently.  
One slight oddness (unless I'm misunderstanding things), and not something that I think should be the cause of the job failure:
This is the covid variable in the study definition.  
https://github.com/opensafely/HEST_OS_study/blob/main/analysis/study_definition.py#L353
It's looking for hospital admissions with codes from the `covid_codelist`, which is [here](https://github.com/opensafely/HEST_OS_study/blob/main/codelists/opensafely-covid-identification.csv), and contains only 2 codes.  

My understanding is that in the output for that column, you'd expect to see ONLY one of those codes, or null.  The generated output from the failed job has lots of different codes in it, not just the 2 from the codelist.

**Comment by evansd** [comment]
Ah, that's a good observation but I think that is explainable. It's returning the `primary_diagnosis` but the codelist is matching on `diagnoses` more generally. So it's finding people hospitalised _with_covid, even though that might not be the primary reason for their hospitalisation.

**Comment by rebkwok** [comment]
Ah, OK.  That makes sense, thanks

**Comment by rebkwok** [comment]
It definitely feels like the issue is in the R script now

**Comment by rebkwok** [comment]
Reporting results of sleuthing:

1) `descriptive_tables` Job is failing intermittently when run locally due to finding no data in the `covid_admission_primary_diagnosis` column (also others, including ethnicity, learning disability)

2) Job also failed on TPP on 24th Nov, with the logs for the descriptive_tables action saying that there's no data for covid_admission_primary_diagnosis.  However, when I went to look at the output file generated by generate_study_population, it does have data in that column.  The logs also say that it identified the covid_admission_primary_diagnosis column as col_logical()

R's `read_csv` function guesses at the type of columns it reads in, on the basis of the first 1000 rows.  If it guesses wrong, it appears to fail to import any data.  In the case on TPP, it has interpreted the column as col_logical(), when it should be col_character().  I was able to reproduce this locally by generating a population with 10,000 rows, and removing any  covid_admission_primary_diagnosis values in the first 1000 rows.  Despite having valid codes in the column, the R script failed to read them.

The incidence of `covid_admission_primary_diagnosis` for dummy data is 0.25; with a large generated population, occasionally the first 1000 rows will be null, resulting in the misinterpretation of the column.

This can be solved by explicitly telling `read_csv` what column types to expect:
```
input <- read_csv(
  here::here("output", "input.csv")
  col_types=cols(.default="?", covid_admission_primary_diagnosis = col_character())  
)
```
(the default here tells it to guess for any unnamed columns.  Any potentially ambiguous columns would need to be specified there)

Alternatively, specifying the output of `generate_study_population` as feather, and reading in a feather file instead of csv appears to deal with the sparse data better:
```
input <- arrow::read_feather(here::here("output", "input.feather"))
```
